### PR TITLE
Change TargetFrameworkVersion to 4.5.2 for Workspaces layer and above

### PR DESCRIPTION
**Pull request by pharring**
Addresses issue #628.
Note that I removed the setting of TargetFrameworkVersion in VSL.Settings.targets. Each project now has to specify the target framework. The compiler layer stays at 4.5 (eventually to move to CoreCLR). 
Note that there are corresponding changes in the Closed tree. I'll send a CodeFlow for those.

### CODE REVIEW

**Comment by pharring** [comment]
@jasonmalinowski please review

**Comment by jaredpar** [comment]
Why does the Visual Studio layer have a 4.5.2 dependency?  Visual Studio itself depends on 4.6 so it would make sense to move them up to that layer as well.

File: src/Workspaces/Core/Portable/Utilities/AsyncSemaphore.cs [hunk 1]
```diff
@@ -2,145 +2,17 @@
...
 using System;
 using System.Threading;
-using System.Threading.Tasks;
 
 namespace Roslyn.Utilities
 {
     /// <summary>
-    /// A semaphore that supports a WaitAsync() operation. Unlike SemaphoreSlim, WaitAsync
-    /// does not have bugs where the semaphore can be acquired but WaitAsync returns cancelled.
+    /// Container for a <see cref="SemaphoreSlim"/> factory.
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Utilities/AsyncSemaphore.cs]: Should this be renamed then to SemaphoreSlimFactory?

**Comment by jasonmalinowski** [comment]
:+1:

My only complaint is you got the honor of deleting AsyncSemaphore...

**Comment by pharring** [comment]
@jaredpar I'm getting there. I've already got most of the 4.6 change done on my machine (interrupted by a build break this morning). Issue #628 was cautious on 4.6. Note that this activates the RS0007 analyzer about using Array.Empty<T>. We have several dozen "violations" in our code-base that I have to fix.

**Comment by jaredpar** [comment]
@pharring ah okay so this is the first in a set of changes.  :+1:

File: src/Workspaces/Core/Portable/Utilities/AsyncSemaphore.cs (hunk 1 above)
Comment (pharring) [inline src/Workspaces/Core/Portable/Utilities/AsyncSemaphore.cs]: Agreed. Will update.

**Comment by jasonmalinowski** [comment]
There's no reason the VS layer couldn't be moved to 4.6, but I don't think it gets us any benefit, really. There's benefits for moving the lower layers to 4.6 but those other cautions apply.

**Comment by jaredpar** [comment]
@jasonmalinowski the reason I asked is that we are going to have to do this in the coming weeks.  The core compiler is moving to CoreFx targetting which will effectively make it 4.6.  The VS layer, as a dependency of the compiler, will need to move to 4.6 as well.  If we are already making the change to move all of VS may as well go ahead and get the move to 4.6 over with.

**Comment by jasonmalinowski** [comment]
Yep, and when that _actually happens_, we can move it to 4.6. A search/replace across the project file isn't hard!

**Comment by jasonmalinowski** [comment]
I like to call it "being realistic".

**Comment by pharring** [comment]
@jaredpar @jasonmalinowski There you go. 4.6 now for Visual Studio layers.

**Comment by jaredpar** [comment]
LGTM!

**Comment by pharring** [comment]
Fixes #628

**Pull request merged**